gunicorn/http/message.py

```diff
@@ -324,6 +330,47 @@ def parse_proxy_protocol(self, line):
             "proxy_port": d_port
         }
 
+    def parse_proxy_protocol_v2(self, unreader, buffer):
+        data = self.read_bytes(unreader, buffer, 4)
+        ver_cmd, fam, length = struct.unpack("!BBH", data)
+        if ver_cmd & 0xF0 != 0x20:
+            raise InvalidProxyHeader("invalid version %r" % data)
+        if ver_cmd & 0xF not in {0x00, 0x01}:
+            raise InvalidProxyHeader("unsupported command %r" % data)
+
+        body = self.read_bytes(unreader, buffer, length)
+        if ver_cmd & 0xF == 0x00:
+            return  # LOCAL command, do not change source addr
+
+        if fam == 0x11:  # TCPv4
+            proto = "TCP4"
+            fmt = "!IIHH"
+            ip_class = ipaddress.IPv4Address
+        elif fam == 0x21:  # TCPv6
+            proto = "TCP6"
+            fmt = "!16s16sHH"
+            ip_class = ipaddress.IPv6Address
+        else:
+            raise InvalidProxyHeader("unsupported protocol %r" % body)
+
+        try:
+            s_addr, d_addr, s_port, d_port = struct.unpack(fmt, body)
+        except struct.error as e:
+            raise InvalidProxyHeader("cannot unpack %r: %s" % (body, e))
+        s_addr = str(ip_class(s_addr))
```

Comment:
Why is it needed? - converting `str -> ipaddress -> str`. Why just not use `str` representation as is?

If it is needed for validation then where is the error handling?

Reply:
@apollo13 ?

Reply:
`s_addr` is never really a string but either an 4-byte integer (ipv4 case) or a 16-byte integer (ipv6 case). Those are "narrow" binary representations of the addresses and not directly usable. So this code is turning the address into something human readable.

As for validation errors: How much do we want to trust the trusted proxies to send the correct data? :D